r/R/singlecell_utils.R

```diff
@@ -68,4 +68,102 @@ marker_features <- function(mat, groups, method="wilcoxon") {
+#' @param threads (integer) Number of threads to use.
+#' @return 
+#'  - If `method` is length `1`, returns a matrix of shape `(features x groups)`.
+#'  - If `method` is greater than length `1`, returns A list of matrices with each matrix representing a pseudobulk matrix with a different aggregation method.
+#' Each matrix is of shape `(features x groups)`, and names are one of `nonzeros`, `sum`, `mean`, `var`.
+#' @details The stats are ordered by complexity: nonzero, sum, mean, then variance. All
+#' less complex stats are calculated in the process of calculating a more complicated stat.
+#' That means it will not take additional time to calculate `mean` if `var` is also requested.
+#' @inheritParams marker_features
+#' @export
+pseudobulk_matrix <- function(mat, cell_groups, method = "sum", clip_values = 1, threads = 1L) {
+  assert_is(mat, "IterableMatrix")
+  assert_is(cell_groups, c("factor", "character", "numeric"))
+  cell_groups <- as.factor(cell_groups)
+  assert_is(method, "character")
+  methods <- c("var", "mean", "sum", "nonzeros")
+  method <- match.arg(method, choices = methods, several.ok = TRUE)
+  assert_is(clip_values, "numeric")
+  assert_is(threads, "integer")
+  # if multiple methods are provided, only need to pass in the top method as it will also calculate the less complex stats
+  top_method <- intersect(methods, method)[1]
+  top_method_number <- match(top_method, rev(methods)) - 1
+  if (clip_values != 1) {
+    quantile_values <- colQuantiles(mat, clip_values)
+    # if all values are 0, skip
+    mat <- min_by_col(mat, quantile_values)
+  }
+  iter <- iterate_matrix(parallel_split(mat, threads, threads*4))
+  cell_groups <- as.factor(cell_groups)
+  res <- pseudobulk_matrix_cpp(iter, cell_groups = as.integer(cell_groups), method = top_method_number, transpose = mat@transpose)
+  # if res is a single matrix, return with colnames and rownames
+  if (length(method) == 1) {
+    colnames(res[[method]]) <- levels(cell_groups)
+    rownames(res[[method]]) <- rownames(mat)
+    return(res[[method]])
+  }
+  # give colnames and rownames for each matrix in res, which is a named list
+  for (res_slot in names(res)) {
+    if ((length(res[[res_slot]]) == 0) || !(res_slot %in% method)) {
+      res[[res_slot]] <- NULL
+    } else {
+      colnames(res[[res_slot]]) <- levels(cell_groups)
+      rownames(res[[res_slot]]) <- rownames(mat)
+    }
+  }
+  return(res)
+}
+
+
+#' Find the nth quantile value of each column in a matrix.
+#' @param x IterableMatrix object or a matrix-like object.
+#' @param probs (Numeric) Quantile value(s) to be computed, between 0 and 1.
+#' @return Numeric vector of quantile values for each column.
+#' @describeIn IterableMatrix-methods Calculate colQuantiles (replacement for `matrixStats::colQuantiles`)
+#' @export
+colQuantiles <- function(x, probs = 0.99, ...) UseMethod("colQuantiles")
+#' @export
+colQuantiles.default <- function(x, probs = 0.99, ..., na.rm = FALSE, type = 7) {
```

Comment:
I've noticed that all the functions that try to utilize the S3 dispatching `colMaxs()` `colVars()` etc all have non-related information in them.  We should discuss how to fix that!